Added UART command and checkbox for sending configuration to NVM.

Joarfo opened the pull request:
Added checkbox in slow control tab that will be used to decide whether to send the config to NVM or directly to cfg_ram. Added UART command for writing to NVM.

Joarfo committed:
Added UART command and checkbox for sending configuration to NVM.

tstana commented on CitirocUI/Citiroc.cs:
```diff
@@ -1924,5 +1922,10 @@ private void Citiroc_FormClosing(object sender, FormClosingEventArgs e)
             }
             catch { /* Blindly close... */}
         }
+
+        private void label_weerocDotCom_Click(object sender, EventArgs e)
+        {
+
+        }
```
This was probably added "automagically" by VS. Suggest you remove it.

tstana commented on CitirocUI/Citiroc.cs:
```diff
@@ -504,8 +504,6 @@ private void comboBox_triggerPreset_SelectedIndexChanged(object sender, EventArg
             }
         }
 
-
-
```
Thanks for removing these lines (if it was indeed you and not Visual Studio this time actually being smart...)! :)

tstana commented on CitirocUI/SlowControl.cs:
```diff
@@ -786,6 +793,11 @@ private void checkBox_bypassPd_CheckedChanged(object sender, EventArgs e)
 
+        private void checkBox_sendToNVM_CheckedChanged(object sender, EventArgs e)
+        {
+            sc_sendToNVM = SC.Switch((CheckBox)sender, button_sendSC);
```
This checkbox you added should have nothing to do with the actual slow control configuration sent to the ASIC. That is what is configured through all of the controls already existing within the Slow Control tab. Rather, this new checkbox you added should offer the user of CitirocUI the capability of selecting which configuration it is they want to send to the NVM. This user selection should make its way to byte 143 in each ASIC configuration, which is actually used by the MSS to store to the proper address in NVM.

tstana replied:
@Joarfo I'm sorry if my comment was unclear, but I quite liked the checkbox you added. What I meant by that comment is that you should not make a call to `SC.Switch()` from here, as this 144-th byte does not make its way to the ASIC, so it doesn't _actually_ belong to the Slow Control configuration sent to the ASIC, which is what the `SC` class handles.

The checkbox and this `checkBox_sendToNVM_CheckedChanged()` method should still exist, it is just that the method should only prepare the 144-th byte to be sent to the ASIC when the **Send SC** is pressed.

Joarfo replied:
Ok, I completely misunderstood your first comment as in the checkbox was terrible and should be removed, which is why I made these changes. Now, in order to not make any more bad changes; how do you want the interface to look? The checkbox exists and when checked it alters the **Send SC** button to be a **Store NVM**? Similarly, should the **Load SC** be changed to **Select NVM**?

tstana replied:
No, the checkbox and the implementation were very good as they were. Unfortunately, the subsequent changes you made were off the mark.

As I mentioned, I think it's best if you start over with a new branch from `master` and a new PR to merge that new branch. You can keep this branch and PR in the meantime, we will close the PR without merging later on.

For the new branch, you should re-add the checkbox. The checkbox would not turn **Send SC** into **Store NVM**, it would rather only change its behavior slightly. Your `NVMVConfigNoSelect` form would be brought up just as it was before, where the user selects the config to be stored (the contents of the 144-th byte appended to the Slow Control configuration). Click **OK** in your form, which calls the `sendSC()` method to send the configuration to Proto-CUBES. You would still need to change the contents of `sendSC()` to send the new command, but not its interface -- so you don't need to make changes in all the files where `sendSC()` is called from. You could use the setting passed from the `NVMVConfigNoSelect` form as a sort of signal to `sendSC()`; this would cause it to use the new command in `ProtoCubesSerial`, which sends the config to NVM.

Remember to hide the new checkbox when the _Citiroc Evaluation Kit_ is selected via the drop-down in the **Connect** tab.

Finally, if it's not too much work, I'd also change the name of `NVMVConfigNoSelect` to `NVMVConfigNumSelect`, but that is obviously not prioritary.

Let me know if something is unclear, I'd be happy to offer some insight.

Joarfo replied:
Ok I think I am clear as to which changes should be made and will fix them tomorrow. 

I am wondering about the functionality for selecting an already loaded configuration on the NVM, should that be implemented on the **Load SC** button, when the checkbox is checked?

tstana reviewed (changes requested):
@Joarfo it looks like you are on the right path, but some more work is still needed. Check first the comments I have added here. Then, in `ProtoCubesSerial.cs` you will likely need to add a case statement within the `SendData()` method that covers the `SendCitiConfNVM` command.

Joarfo committed:
Added commands and implementation for Selecting stored ASIC config and
uploading config

tstana commented on CitirocUI/SlowControl.cs:
```diff
@@ -541,9 +541,45 @@ void slowControl_DragDrop(object sender, DragEventArgs e)
+                        {
+                            configNo = Convert.ToString(Convert.ToInt32(configNo, 10), 2);
+                            configNo = configNo.PadLeft(8, '0');
+                            string strSC = getSC();
```
No need to redefine `strSC` here, you already have it as a member in the class. It should just be possible (I think) to do `strSC += configNo` as you do below, without this `getSC()`.

tstana commented on CitirocUI/SlowControl.cs:
```diff
@@ -541,9 +541,45 @@ void slowControl_DragDrop(object sender, DragEventArgs e)
+                {
+                    try
+                    {   
+                        if (Convert.ToInt32(configNo, 10)>255)
```
Instead of converting to an Int32, you could add to the form a simple feature to truncate everything above 255 and then do `Convert.ToByte(configNo)`. You can then remove the if/else statement within the try statement and simplify the code.

tstana commented on CitirocUI/SlowControl.cs:
```diff
@@ -1410,27 +1446,118 @@ private void button_saveSC_Click(object sender, EventArgs e)
-            ScLoadDialog.Title = "Specify Data file";
-            //DataLDialog.Filter = "TxT files|*.txt";
-            ScLoadDialog.RestoreDirectory = true;
+            //Initialize string for input of config number selection
+            string configNo = "";
+            if (sc_sendToNVM == 1)
```
I see no need to ask the user for input on `loadSC()`. When the user clicks this button, they should just be brought to a dialog that allows them to select different configurations in the default format of CitirocUI, i.e., without the CUBES addition of the extra ID byte. It is safe to remove the if statement here.

tstana reviewed (changes requested):
Sorry, @Joarfo, but more work is needed. Apart from my in-line comments, please also address the following:

- the comments I gave you on my last review
- when the instrument selected via the combo box in the **Connect** pane changes to the _Citiroc Evaluation Kit_, the CUBES-specific controls that allow for sending the configuration to the NVM should be hidden from the user. We need to have the capability of using the CitirocUI with both CUBES and the Citiroc Eval Kit, without having confusing controls that are not used appear in the case of the Citiroc Development Kit. The method `comboBox_SelectConnection_SelectedIndexChanged()` is where the hiding of CUBES-specific controls is done.

Joarfo committed:
Removed SendToNVM checkbox and added a store and
select configuration button for the NVM. Added file
for handling text inputs in message box. Added
command as an argument for SendSC.

Joarfo committed:
Update SlowControl.cs

Corrected a mistyping

Joarfo committed:
Removed test-pop-ups and changed button text from "store" to "store NVM"

tstana reviewed (commented):
@Joarfo unfortunately the changes that you made in this commit do not make for maintainable code. In particular, changing the interface of the `sendSC()` method is not good practice for the reasons outlined in the comment to the `sendSC()` method.

I suggest we close this PR without merging and start over with a new PR based off the current `master`.

In this new PR, we would re-add the checkbox you had several commits ago and only change the contents of the `sendSC()` method, or add a new function to specifically send the SC only to the NVM for CUBES. At any rate, changing the interface of `sendSC()` -- or any method in use -- should be avoided to any extent possible.

Also to be kept in mind: The `NVMVConfigNoSelect` form can probably be passed over to the new PR.

tstana commented:
PR #16 was opened to restart work on these changes. This PR will be closed (without merging).
